llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -3261,6 +3288,15 @@ const SCEV *ScalarEvolution::getMulExpr(SmallVectorImpl<const SCEV *> &Ops,
       SCEV::NoWrapFlags Flags =
           AddRec->getNoWrapFlags(ComputeFlags({Scale, AddRec}));
 
+      // Preserve flags for positive constant Scale.
+      if (auto *SC = dyn_cast<SCEVConstant>(Scale))
+        if (SC->getAPInt().isStrictlyPositive()) {
+          if (hasFlags(OrigFlags, SCEV::FlagNSW))
+            Flags = setFlags(Flags, SCEV::FlagNSW);
+          if (hasFlags(OrigFlags, SCEV::FlagNUW))
+            Flags = setFlags(Flags, SCEV::FlagNUW);
+        }
```

Comment:
If I understand correctly, the claim you're making here is that `scale *<nuw> {start,+,step}` is `{scale*start,+,scale*step}<nuw>` -- that is, this only depends on the flags on the multiply, not the addrec (unlike the intersection taken directly above). This seems trivially incorrect in the case where `scale=1`, in which case the multiplication is always nuw and you gain zero information on the addrec.

But even for non-trivial cases, let's take `2 *<nuw> {1,+,-1}` where the addrec takes values `{1,0}` The multiply is nuw, but the addrec `{2,+,-2}` isn't.